angularSTEP/src/app/recipe-page/recipe-page.component.ts

```diff
@@ -13,17 +16,59 @@ import {RecipeConverter} from '../recipe-converter';
 export class RecipePageComponent {
   id: string | null = this.route.snapshot.paramMap.get('id');
   pageRecipe!: Recipe;
+  signedIn: boolean = false;
+  currentRating: number = 0;
+  currentRatings: object = {};
 
   constructor(private db: AngularFirestore,
-              private route: ActivatedRoute,) {
-                this.setRecipeData();
+              private route: ActivatedRoute, private fAuth: AngularFireAuth) {}
+
+  ngOnInit() {
+    this.setRecipeData();
   }
 
   async setRecipeData() {
     const postRecipe = await this.db.doc('/recipes/' + this.id + '/').ref.withConverter(new RecipeConverter().recipeConverter).get();
     const recipeClassData = postRecipe.data();
     if (recipeClassData) {
       this.pageRecipe = recipeClassData;
+      if(this.pageRecipe.ratings){
+        this.currentRatings = this.pageRecipe.ratings;
+      }
     }
+    this.fAuth.onAuthStateChanged(async user => {
+      if(user) {
+        this.signedIn = true;
+        if(this.currentRatings && this.currentRatings.hasOwnProperty(user.uid)){
+          this.currentRating = this.currentRatings[user.uid];
+        } else {
+          this.currentRating = 0;
+        }
+      } else {
+        this.signedIn = false;
+      }
+    })
   } 
+
+  updateRating(rating: number){
+    this.fAuth.currentUser.then(user => {
+      if(this.signedIn && user) {
+        this.currentRatings[user.uid] = rating;
+        this.db
+            .collection('recipes')
+            .doc(this.id)
+            .ref.withConverter(new RecipeConverter().recipeConverter)
+            .update({ratings: this.currentRatings, averageRating: this.getAverageRating()})
+      }
+    })
+  }
+
+  getAverageRating() {
+    let sum = 0;
+    for(let key of Object.keys(this.currentRatings)){
+      sum += this.currentRatings[key];
+    }
+    this.pageRecipe.averageRating = Math.round((sum / Object.keys(this.currentRatings).length)*2)/2;
```

Comment:
Did you mean to round-off to the nearest half-star, or did you intend to use Math.floor? If the former, please add a comment explaining why.

Reply:
Oh, yes I should add a comment. I actually did mean to round to the nearest half star just to give a little more different variation in the ratings.